sim/midas/src/main/scala/midas/models/sram/BankedAsyncMemModel.scala

```diff
@@ -0,0 +1,138 @@
+package midas.models.sram
+
+import chisel3._
+import chisel3.util.{PriorityEncoderOH, DecoupledIO, log2Ceil}
+
+object BankedAsyncMemChiselModel {
+  /**
+    * NOTE: Enabling read skips can change simulation results for buggy targets that use undefined read data
```

Comment:
This implementation is sensitive to the order of arrival of write tokens in terms of which write wins for write-write collisions (a separate distinction from that in the note). As long as all the write ports follow the same "model path" (which will generally just go from one model containing the CPU to a regfile model) from a register to the write port, this will not introduce any non-determinism. However, it is possible to construct situations where it would. A potential resolution would be to add an assertion that asserts a lack of write-write collisions and error with "warning, your setup can produce non-determinism" if they exist.

Reply:
Ah, right. That's subtle. You could also check that the write-token order never changes from cycle to cycle. since as you point out, it's probably going to be static in 99.9% of foreseeable cases. I think starting with a write collision check is reasonable.

Reply:
> I think this is the PR @davidbiancolin mentioned would improve the FMR if I try to start using the MCRams optimization, please correct me if I've wandered into a different PR. @albert-magyar, if I want to try this out, I need to do something to make the `LabelSRAMMacros` pass use this model, right? I'm starting to look at that code but if you could give me a pointer on how to try this out, I'd really appreciate it!

Sure -- basically, all you need to do is change [this line](https://github.com/firesim/firesim/blob/9598b7f4448e693fe9f3b18689a55503a7ea08c8/sim/midas/src/main/scala/midas/passes/fame/EmitAndWrapRAMModels.scala#L195) to use `BankedAsyncMemChiselModel` instead of `AsyncMemChiselModel`. They take the same parameters, aside from some optional ones for the banked version that have sensible defaults and control the configuration of the underlying host memory.

(Sorry for the late reply; perhaps you already found this out.)

Reply:
Something else to consider: there is no guarantee that multi-write-port memories will be deterministic in the *target* under write collisions, since this is undefined under FIRRTL. This can affect both FPGA prototypes and unoptimized models that contain these memories. In practice, combinational-read memories have deterministic write collisions, but this is relying on implementation-defined behavior. However, anything that uses BRAMs to implement 2-RW synchronous-read memories (including the new "FPGA backend") will result in the non-deterministic write collision behavior of true dual-port BRAMs being exposed to unoptimized simulators of target designs that generate write collisions to such memories.